platform/include/logging_stack.h

```diff
@@ -0,0 +1,93 @@
+/*
+ * Logging Stack V1.0.0
+ * Copyright (C) 2020 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file logging_stack.h
+ * @brief Reference implementation of Logging stack as a header-only library.
+ */
+
+#ifndef LOGGING_STACK_
+#define LOGGING_STACK_
+
+/* Include header for logging level macros. */
+#include "logging_levels.h"
+
+/* Standard Include. */
+#include <stdio.h>
+
+/* Metadata information to prepend to every log message. */
+#define LOG_METADATA_FORMAT_ARGS    ( "[%s:%d] ", __FILE__, __LINE__ )
```

Comment:
What does LOG_METADATA_FORMAT_ARGS and SdkLogWithNewLine serve? Can this simply not be:

```c
  #define LogError( message )    SdkLog( ( "[ERROR] [%s] [%s: %d] ", LIBRARY_LOG_NAME, __FILE__, __LINE__ ) );  SdkLog( message ); SdkLog( ( "\r\n" ) );
```

Also, do we really need the call for appending a new line here or can it be left to the call sites? The reason I am pointing this out is because on FreeRTOS platforms our logging implementation serializes the access to UART by posting all the logging messages to a queue which is drained by a dedicated logging task. In this case we will end up posting 3 messages to the queue for every log and it may result in dropped messages if the logging task is not able to keep up.

Reply:
The intention of the `LOG_METADATA_FORMAT_ARGS ` macro is to provide a single location of changing the metadata information. (It is primarily for convenience of changing the metadata and better visibility).
Yes, for the new line, we avoid the `SdkLogWithNewline` macro.